lib/std/parker.zig

```diff
@@ -0,0 +1,320 @@
+    .linux => if (builtin.link_libc) PosixParker else LinuxParker,
+    .windows => WindowsParker,
+    else => SpinParker,
+};
+
+const SpinParker = struct {
+    pub fn init() SpinParker {
+        return SpinParker{};
+    }
+    pub fn deinit(self: *SpinParker) void {}
+
+    pub fn unpark(self: *SpinParker, ptr: *const u32) void {}
+
+    pub fn park(self: *SpinParker, ptr: *const u32, expected: u32) void {
+        var backoff = SpinLock.Backoff.init();
+        while (@atomicLoad(u32, ptr, .Acquire) == expected)
+            backoff.yield();
+    }
+};
+
+const LinuxParker = struct {
+    pub fn init() LinuxParker {
+        return LinuxParker{};
+    }
+    pub fn deinit(self: *LinuxParker) void {}
+
+    pub fn unpark(self: *LinuxParker, ptr: *const u32) void {
+        const rc = linux.futex_wake(@ptrCast(*const i32, ptr), linux.FUTEX_WAKE | linux.FUTEX_PRIVATE_FLAG, 1);
+        assert(linux.getErrno(rc) == 0);
+    }
+
+    pub fn park(self: *LinuxParker, ptr: *const u32, expected: u32) void {
+        const value = @intCast(i32, expected);
+        while (@atomicLoad(u32, ptr, .Acquire) == expected) {
+            const rc = linux.futex_wait(@ptrCast(*const i32, ptr), linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, value, null);
+            switch (linux.getErrno(rc)) {
+                0, linux.EAGAIN => return,
+                linux.EINTR => continue,
+                linux.EINVAL => unreachable,
+                else => unreachable,
+            }
+        }
+    }
+};
+
+const WindowsParker = struct {
+    waiters: u32,
+
+    pub fn init() WindowsParker {
+        return WindowsParker{ .waiters = 0 };
+    }
+    pub fn deinit(self: *WindowsParker) void {}
+
+    pub fn unpark(self: *WindowsParker, ptr: *const u32) void {
+        const key = @ptrCast(*const c_void, ptr);
+        const handle_ptr = getEventHandlePtr() orelse return;
+
+        var waiting = @atomicLoad(u32, &self.waiters, .Monotonic);
+        while (waiting != 0) {
+            waiting = @cmpxchgWeak(u32, &self.waiters, waiting, waiting - 1, .Acquire, .Monotonic) orelse {
+                const rc = windows.ntdll.NtReleaseKeyedEvent(handle_ptr.*, key, windows.FALSE, null);
+                assert(rc == 0);
+                return;
+            };
+        }
+    }
+
+    pub fn park(self: *WindowsParker, ptr: *const u32, expected: u32) void {
+        var spin = SpinLock.Backoff.init();
+        const key = @ptrCast(*const c_void, ptr);
+        
+        while (@atomicLoad(u32, ptr, .Acquire) == expected) {
+            if (getEventHandlePtr()) |handle_ptr| {
+                _ = @atomicRmw(u32, &self.waiters, .Add, 1, .Release);
+                const rc = windows.ntdll.NtWaitForKeyedEvent(handle_ptr.*, key, windows.FALSE, null);
+                assert(rc == 0);
+            } else {
+                spin.yield();
+            }
+        }
+    }
+
+    var event_handle = std.lazyInit(windows.HANDLE);
+
+    fn getEventHandlePtr() ?*const windows.HANDLE {
```

Comment:
BTW sorry to nitpick but why does this return a `?*const windows.HANDLE` and not `?windows.HANDLE`? At both of your callsites, you always deref this value. I think it might make more sense here to `return event_handle.data` rather than a pointer to the field.

Reply:
Its fine and that indeed does makes more sense. Was only lazy on my part to not deref it in getEventHandle. Out of curiosity is there a style guide or like a "general rules to follow" when PR'ing code?

Reply:
Not really. I try to keep the overhead for contributors low, especially while the std lib is considered unstable. If the fixups are simple, often I'll just do them so the contributor doesn't have to. 

 If such a document would be useful then that content could go in CONTRIBUTING.md. it would mostly be advice on how to get one's PR to be more likely to get merged quickly.